spec/unit/pops/types/p_type_set_type_spec.rb

```diff
@@ -0,0 +1,424 @@
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(Puppet::Error,
+                /entry 'types' expected a Hash value, got Array/)
+            end
+
+            it 'contains values that are not types' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                types => {
+                  Car => 'brum'
+                }
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(Puppet::Error,
+                /The expression <'brum'> is not a valid type specification/)
+            end
+
+            it 'contains keys that are not SimpleNames' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                types => {
+                  car => Integer
+                }
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                /key of entry 'car' expected a match for Pattern\[\/\[A-Z\]\\w\*\/\], got 'car'/)
+            end
+          end
+
+          context 'the references hash' do
+            it 'is empty' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                references => {}
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                /entry 'references' expected size to be at least 1, got 0/)
+            end
+
+            it 'is not a hash' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                references => []
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                /entry 'references' expected a Hash value, got Array/)
+            end
+
+            it 'contains something other than reference initialization maps' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                references => {Ref => 2}
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(TypeAssertionError,
+                /entry 'references' entry 'Ref' expected a Struct value, got Integer/)
+            end
+
+            it 'contains several initialization that refers to the same TypeSet' do
+              ts = <<-OBJECT
+                pcore_version => '1.0.0',
+                version => '1.0.0',
+                references => {
+                  A => { name => 'Vehicle::Cars', version_range => '1.x' },
+                  V => { name => 'Vehicle::Cars', version_range => '1.x' },
+                }
+              OBJECT
+              expect { parse_type_set('MySet', ts) }.to raise_error(ArgumentError,
+                /references TypeSet 'http:\/\/puppet\.com\/2016\.1\/runtime\/Vehicle::Cars'/)
```

Comment:
this error message seems a bit off - how would the user know it means the same set more than once?